### PR TITLE
Add mechanism for executing work in the background

### PR DESCRIPTION
**Pull request by gunndabad**
Usage:
```C#
IBackgroundWorkScheduler scheduler;
await scheduler.Schedule(RunMyBackgroundWork, state: new MyState());

static Task RunMyBackgroundWork(object state, IServiceProvider serviceProvider, CancellationToken cancellationToken)
{
    MyState stronglyTypedState = (MyState)state;
    var dep = serviceProvider.GetRequiredService<ISomeDependency>();
    await dep.DoTheThing(state.Foo);
}
```

### CODE REVIEW

File: tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+                cancelledSignal.Wait(100);
+                Assert.True(ct.IsCancellationRequested);
+
+                executedSignal.Set();
+                return Task.CompletedTask;
+            };
+
+            // Act
+            await scheduler.StartAsync(cancellationToken: default);
```
Comment (jameslawesynergy) [inline tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs:115]: Unsure as to why the complexity is needed here? Just capture the passed `CancellationToken` and `Assert.Equals`?

File: tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            using var executingSignal = new ManualResetEventSlim(false);
+            using var cancelledSignal = new ManualResetEventSlim(false);
+            using var executedSignal = new ManualResetEventSlim(false);
+
+            WorkItem workItem = (state, sp, ct) =>
+            {
+                executingSignal.Set();
+
+                cancelledSignal.Wait(100);
+                Assert.True(ct.IsCancellationRequested);
```
Comment (jameslawesynergy) [inline tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs:108]: As above, I think a false assertion would be caught by the `Execute()` try catch.

File: tests/Dfc.CourseDirectory.Testing/ExecuteImmediatelyBackgroundWorkScheduler.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Threading.Tasks;
+using Dfc.CourseDirectory.Core.BackgroundWorkers;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Dfc.CourseDirectory.Testing
+{
+    public class ExecuteImmediatelyBackgroundWorkScheduler : IBackgroundWorkScheduler
```
Comment (jameslawesynergy) [inline tests/Dfc.CourseDirectory.Testing/ExecuteImmediatelyBackgroundWorkScheduler.cs:7]: I can't see where this is being used...

File: src/Dfc.CourseDirectory.Core/BackgroundWorkers/QueueBackgroundWorkScheduler.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                await foreach (var entry in _work.Reader.ReadAllAsync(stoppingToken))
+                {
+                    using var scope = _serviceScopeFactory.CreateScope();
+                    var scopeServices = scope.ServiceProvider;
+
+                    try
+                    {
+                        await entry.WorkItem(entry.State, scopeServices, stoppingToken);
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Core/BackgroundWorkers/QueueBackgroundWorkScheduler.cs:66]: The task itself can choose whether or not to observe the token, if running to completion is required

File: tests/Dfc.CourseDirectory.Testing/ExecuteImmediatelyBackgroundWorkScheduler.cs (hunk 3 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.Testing/ExecuteImmediatelyBackgroundWorkScheduler.cs:7]: I needed this as part of next PR for testing but it felt like it made more sense in this PR

File: tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs (hunk 1 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs:128]: Additional `executingSignal` is required to know that the item has started executing before the cancellation is invoked

File: tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                }
+            };
+
+            // Act
+            await scheduler.StartAsync(cancellationToken: default);
+            await scheduler.Schedule(workItem);
+            executedSignal.Wait(100);
+
+            // Assert
+            await scheduler.StopAsync(cancellationToken: default);
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.Core.Tests/BackgroundWorkersTests/QueueBackgroundWorkSchedulerTests.cs:71]: Fixed now

**Pull request merged**